c2rust-analyze/src/main.rs

```diff
@@ -544,6 +546,110 @@ fn run(tcx: TyCtxt) {
+                let ldid = match func_def_path_hash_to_ldid.get(&def_path_hash) {
+                    Some(&x) => x,
+                    None => {
+                        eprintln!(
+                            "pdg: unknown DefPathHash {:?} for function {:?}",
+                            n.function.id, n.function.name
+                        );
+                        continue;
+                    }
+                };
+                let info = func_info.get_mut(&ldid).unwrap();
+                let ldid_const = WithOptConstParam::unknown(ldid);
+                let mir = tcx.mir_built(ldid_const);
+                let mir = mir.borrow();
+                let acx = gacx.function_context_with_data(&mir, info.acx_data.take());
+                let mut asn = gasn.and(&mut info.lasn);
+
+                let dest_pl = match n.dest.as_ref() {
+                    Some(x) => x,
+                    None => {
+                        info.acx_data.set(acx.into_data());
+                        continue;
+                    }
+                };
+                if dest_pl.projection.len() > 0 {
+                    info.acx_data.set(acx.into_data());
+                    continue;
+                }
+                let dest = dest_pl.local;
+                let dest = Local::from_u32(dest.index);
+
+                let ptr = match acx.ptr_of(dest) {
+                    Some(x) => x,
+                    None => {
+                        eprintln!(
+                            "pdg: {}: local {:?} appears as dest, but has no PointerId",
+                            n.function.name, dest
+                        );
+                        info.acx_data.set(acx.into_data());
+                        continue;
+                    }
+                };
+
+                let node_info = match n.info.as_ref() {
+                    Some(x) => x,
+                    None => {
+                        eprintln!(
+                            "pdg: {}: node with dest {:?} is missing NodeInfo",
+                            n.function.name, dest
+                        );
+                        info.acx_data.set(acx.into_data());
+                        continue;
+                    }
+                };
+
+                let old_perms = asn.perms()[ptr];
+                let mut perms = old_perms;
+                if node_info.flows_to.load.is_some() {
+                    perms.insert(PermissionSet::READ);
+                }
+                if node_info.flows_to.store.is_some() {
+                    perms.insert(PermissionSet::WRITE);
+                }
+                if node_info.flows_to.pos_offset.is_some() {
+                    perms.insert(PermissionSet::OFFSET_ADD | PermissionSet::OFFSET_SUB);
+                }
+                if node_info.flows_to.neg_offset.is_some() {
+                    perms.insert(PermissionSet::OFFSET_ADD | PermissionSet::OFFSET_SUB);
+                }
```

Comment:
Shouldn't `pos_offset` correspond to `OFFSET_ADD` and `neg_offset` to `OFFSET_SUB`?

Reply:
CC @spernsteiner?

Reply:
That sounds right - not sure why I did it this way.

Reply:
Changed in db5ef66.